{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Imports\n",
    "Versions <br>\n",
    "python - 3.6.7 <br>\n",
    "Pandas - 0.23.4 <br>\n",
    "Biopython - 1.73 <br>\n",
    "Scipy - 1.2.0 <br>\n",
    "iupred2d - https://iupred2a.elte.hu/download <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import ntpath\n",
    "import datetime\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from scipy import signal\n",
    "from iupred2a import iupred2a\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "# Windows\n",
    "CWD = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Statics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "RESIDUES = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L',\n",
    "            'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "# Kyte & Doolittle {kd} index of hydrophobicity\n",
    "HP = {'A': 1.8, 'R':-4.5, 'N':-3.5, 'D':-3.5, 'C': 2.5,\n",
    "      'Q':-3.5, 'E':-3.5, 'G':-0.4, 'H':-3.2, 'I': 4.5,\n",
    "      'L': 3.8, 'K':-3.9, 'M': 1.9, 'F': 2.8, 'P':-1.6,\n",
    "      'S':-0.8, 'T':-0.7, 'W':-0.9, 'Y':-1.3, 'V': 4.2, 'U': 0.0}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import time\n",
    "class IuPred:\n",
    "    def __init__(self, glob, short, long):\n",
    "        self.glob = glob\n",
    "        self.short = short\n",
    "        self.long = long\n",
    "\n",
    "\n",
    "class HydroPhobicIndex:\n",
    "    def __init__(self, hpilist):\n",
    "        self.hpilist = hpilist\n",
    "\n",
    "\n",
    "class MakeMatrix:\n",
    "    def __init__(self, dbfasta):  \n",
    "        self.df = pd.DataFrame()\n",
    "        executables = [\n",
    "            'self.fasta2df(dbfasta)',\n",
    "            'self.amino_acid_analysis()',\n",
    "            'self.idr_iupred()',\n",
    "            'self.hydrophobic()',\n",
    "            'self.add_iupred_features()',\n",
    "            'self.add_hydrophobic_features()',\n",
    "            'self.add_biochemical_combinations()',\n",
    "            'self.add_lowcomplexity_features()' ,\n",
    "            #'self.add_plaac()'\n",
    "        ]        \n",
    "        for e in executables:\n",
    "            start = time.time()     \n",
    "            exec(e)\n",
    "            end = time.time()        \n",
    "            print(str(round(end - start, 2))+'s '+e)\n",
    "\n",
    "    def fasta2df(self, dbfasta):\n",
    "        rows = list()\n",
    "        with open(dbfasta) as f:\n",
    "            for record in SeqIO.parse(dbfasta, 'fasta'):\n",
    "                seqdict = dict()\n",
    "                seq = str(record.seq)\n",
    "                id = record.description.split('|')                \n",
    "                if id[0] == 'sp':\n",
    "                    uniprot_id = id[1]\n",
    "                    name = id[2].split(' ')[0]\n",
    "                    rows.append([name, uniprot_id, seq])\n",
    "                else:\n",
    "                    uniprot_id = id[0]\n",
    "                    name = id[2].split(' ')[0]\n",
    "                    rows.append([name, uniprot_id, seq])                    \n",
    "        self.df = pd.DataFrame(rows, columns=['protein_name', 'uniprot_id', 'sequence'])\n",
    "        print(len(self.df))\n",
    "\n",
    "    def set_identifiers(self, fastas):\n",
    "        self.df['llps'] = 0\n",
    "        for i in fastas:\n",
    "            fname = ntpath.basename(i).split('.')[0]\n",
    "            self.df[fname] = 0\n",
    "            with open(i) as f:\n",
    "                for record in SeqIO.parse(f, 'fasta'):\n",
    "                    id_split = record.id.split('|')\n",
    "                    protid = (id_split[1])\n",
    "                    self.df.loc[self.df['uniprot_id'] == protid, fname] = 1\n",
    "                    self.df.loc[self.df['uniprot_id'] == protid, 'llps'] = 1\n",
    "\n",
    "    def idr_iupred(self):\n",
    "        self.df['iupred'] = object\n",
    "        for index, row in self.df.iterrows():\n",
    "            seq = row['sequence']\n",
    "            glob = iupred2a.iupred(str(seq), 'glob')\n",
    "            short = iupred2a.iupred(str(seq), 'short')\n",
    "            long = iupred2a.iupred(str(seq), 'long')\n",
    "            idrpred = IuPred(glob, short, long)\n",
    "            self.df.at[index, 'iupred'] = idrpred\n",
    "            \n",
    "    def hydrophobic(self):\n",
    "        for index, row in self.df.iterrows():\n",
    "            hpilst = pd.Series(list(row['sequence'])).map(HP).tolist()\n",
    "            self.df.loc[index, 'HydroPhobicIndex'] = HydroPhobicIndex(hpilst)\n",
    "            \n",
    "    def amino_acid_analysis(self):\n",
    "        for res in RESIDUES:\n",
    "            self.df['fraction_'+res] = self.df['sequence'].str.count(res) / self.df['sequence'].str.len()\n",
    "        self.df['length'] = self.df['sequence'].str.len()        \n",
    "        for index, row in self.df.iterrows():\n",
    "            seq = row['sequence']   \n",
    "            seqanalysis = ProteinAnalysis(seq)\n",
    "            acidist = seqanalysis.get_amino_acids_percent() \n",
    "            self.df.loc[index, 'IEP'] = seqanalysis.isoelectric_point()\n",
    "            if 'X' not in seq and 'B' not in seq:\n",
    "                self.df.loc[index, 'molecular_weight'] = seqanalysis.molecular_weight()\n",
    "            if 'U' not in seq and 'X' not in seq and 'B' not in seq:\n",
    "                self.df.loc[index, 'gravy'] = seqanalysis.gravy()\n",
    "          \n",
    "\n",
    "    def add_iupred_features(self):\n",
    "        for index, row in self.df.iterrows():\n",
    "            idr = row['iupred'].glob[0]\n",
    "            self.df.loc[index, 'idr_percetage'] = sum(i > .5 for i in list(idr))\n",
    "            self.df.loc[index, 'idr_50'] = sum(i > .5 for i in list(idr)) / len(str(row['sequence']))\n",
    "            self.df.loc[index, 'idr_60'] = sum(i > .6 for i in list(idr)) / len(str(row['sequence']))\n",
    "            self.df.loc[index, 'idr_70'] = sum(i > .7 for i in list(idr)) / len(str(row['sequence']))\n",
    "            self.df.loc[index, 'idr_80'] = sum(i > .8 for i in list(idr)) / len(str(row['sequence']))\n",
    "            self.df.loc[index, 'idr_90'] = sum(i > .9 for i in list(idr)) / len(str(row['sequence']))\n",
    "\n",
    "    def add_hydrophobic_features(self):\n",
    "        hpi0, hpi1, hpi2, hpi3, hpi4, hpi5 = list(), list(), list(), list(), list(), list() \n",
    "        for index, row in self.df.iterrows():\n",
    "            sw = convolve_signal(row['HydroPhobicIndex'].hpilist, window=30)\n",
    "            hpi0.append(sum(i < -1.5 for i in sw) / len(sw))\n",
    "            # self.df.loc[index, 'hpi_<-1.5_frac'] = hpi\n",
    "            hpi1.append(sum(i < -2.0 for i in sw) / len(sw))\n",
    "            # self.df.loc[index, 'hpi_<-2.0_frac'] = hpi\n",
    "            hpi2.append(sum(i < -2.5 for i in sw) / len(sw))\n",
    "            # self.df.loc[index, 'hpi_<-2.5_frac'] = hpi\n",
    "            hpi3.append(sum(i < -1.5 for i in sw))\n",
    "            # self.df.loc[index, 'hpi_<-1.5'] = hpi\n",
    "            hpi4.append( sum(i < -2.0 for i in sw))\n",
    "            # self.df.loc[index, 'hpi_<-2.0'] = hpi\n",
    "            hpi5.append(sum(i < -2.5 for i in sw))\n",
    "            # self.df.loc[index, 'hpi_<-2.5'] = hpi \n",
    "        self.df['hpi_<-1.5_frac'] = hpi0\n",
    "        self.df['hpi_<-2.0_frac'] = hpi1\n",
    "        self.df['hpi_<-2.5_frac'] = hpi2\n",
    "        self.df['hpi_<-1.5'] = hpi3\n",
    "        self.df['hpi_<-2.0'] = hpi4\n",
    "        self.df['hpi_<-2.5'] = hpi5\n",
    "            \n",
    "\n",
    "    def add_biochemical_combinations(self):\n",
    "        df = self.df\n",
    "        df = df.assign(Asx=df['fraction_D'] + df['fraction_N'])\n",
    "        df = df.assign(Glx=df['fraction_E'] + df['fraction_Q'])\n",
    "        df = df.assign(Xle=df['fraction_I'] + df['fraction_L'])\n",
    "        df = df.assign(Pos_charge=df['fraction_K'] + df['fraction_R'] + df['fraction_H'])\n",
    "        df = df.assign(Neg_charge=df['fraction_D'] + df['fraction_E'])\n",
    "        df = df.assign(Aromatic=df['fraction_F'] + df['fraction_W'] + df['fraction_Y'] + df['fraction_H'])\n",
    "        df = df.assign(Alipatic=df['fraction_V'] + df['fraction_I'] + df['fraction_L'] + df['fraction_M'])\n",
    "        df = df.assign(Small=df['fraction_P'] + df['fraction_G'] + df['fraction_A'] + df['fraction_S'])\n",
    "        df = df.assign(Hydrophilic=(df['fraction_S'] + df['fraction_T'] + df['fraction_H'] + \n",
    "                                    df['fraction_N'] + df['fraction_Q'] + df['fraction_E'] +\n",
    "                                    df['fraction_D'] + df['fraction_K'] + df['fraction_R']))\n",
    "        df = df.assign(Hydrophobic= (df['fraction_V'] + df['fraction_I'] + df['fraction_L'] +\n",
    "                                     df['fraction_F'] + df['fraction_W'] + df['fraction_Y'] +\n",
    "                                     df['fraction_M']))\n",
    "        \n",
    "        # Added in version 2\n",
    "        for dimer in ['GV', 'VG', 'VP', 'PG', 'FG', 'RG', 'GR', 'GG', 'YG', 'GS', 'SG', 'GA', 'GF', 'GD', 'DS']:\n",
    "            self.df[dimer] = self.df['sequence'].str.count(dimer)\n",
    "        df = df.assign(alpha_helix=df['fraction_V'] + df['fraction_I'] + df['fraction_Y'] + df['fraction_F']\n",
    "                      + df['fraction_W'] + df['fraction_L'])\n",
    "        df = df.assign(beta_turn=df['fraction_N'] + df['fraction_P'] + df['fraction_G'] + df['fraction_S'])\n",
    "        df = df.assign(beta_sheet=df['fraction_E'] + df['fraction_M'] + df['fraction_A'] + df['fraction_L'])\n",
    "        #Calculates the aromaticity value of a protein according to Lobry, 1994. \n",
    "        # It is simply the relative frequency of Phe+Trp+Tyr.\n",
    "        df = df.assign(aromaticity=df['fraction_F'] + df['fraction_W'] + df['fraction_Y'])\n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "        self.df = df\n",
    "        del df\n",
    "        \n",
    "    def add_lowcomplexityscore(self):\n",
    "        lcs_window = 20\n",
    "        lcs_cutoff = 7\n",
    "        for index, row in self.df.iterrows():\n",
    "            seq = str(row['sequence'])\n",
    "            if len(seq) > lcs_window+1:\n",
    "                sig = list()\n",
    "                for i in range(len(seq)):\n",
    "                    window = (seq[i: i+lcs_window])\n",
    "                    if len(window) == lcs_window:\n",
    "                        acid_comp = len(list(set(window)))\n",
    "                        sig.append(acid_comp)\n",
    "                score = sum([1 if i <= 7 else 0 for i in sig])\n",
    "                self.df.loc[index, 'lcs_score'] = score\n",
    "                self.df.loc[index, 'lcs_fraction'] = score / len(sig)\n",
    "                \n",
    "                \n",
    "    def add_lowcomplexity_features(self):\n",
    "        n_window = 20\n",
    "        cutoff = 7       \n",
    "        n_halfwindow = int(n_window / 2)        \n",
    "        lcs_lowest_complexity = list()\n",
    "        lcs_scores = list()\n",
    "        lcs_fractions = list()       \n",
    "        for index, row in self.df.iterrows():            \n",
    "            # Determine low complexity scores\n",
    "            seq = str(row['sequence'])\n",
    "            lcs_acids = list()\n",
    "            sig = list()\n",
    "            \n",
    "            # New\n",
    "            lc_bool = [False] * len(seq)\n",
    "            for i in range(len(seq)):\n",
    "                if i < n_halfwindow:\n",
    "                    peptide = seq[:n_window]        \n",
    "                elif i+n_halfwindow > int(len(seq)):\n",
    "                    peptide = seq[-n_window:]        \n",
    "                else:\n",
    "                    peptide = seq[i-n_halfwindow:i+n_halfwindow]       \n",
    "                complexity = (len(set(peptide)))\n",
    "                if complexity <= 7:\n",
    "                    for bool_index in (i-n_halfwindow, i+n_halfwindow):\n",
    "                        try:\n",
    "                            lc_bool[bool_index] = True\n",
    "                        except IndexError:\n",
    "                            pass\n",
    "                    lcs_acids.append(seq[i])\n",
    "                sig.append(complexity)            \n",
    "            # Adding low complexity scores to list\n",
    "            low_complexity_list = pd.DataFrame({'bool':lc_bool, 'acid':list(seq)}, index=None)\n",
    "            lcs_lowest_complexity.append(min(sig))\n",
    "            lcs_scores.append(len(low_complexity_list.loc[low_complexity_list['bool'] == True]))\n",
    "            lcs_fractions.append(len(low_complexity_list.loc[low_complexity_list['bool'] == True]) / len(seq))\n",
    "            low_complexity_list = pd.DataFrame({'bool':lc_bool, 'acid':list(seq)}, index=None)\n",
    "            if len(lcs_acids) >= n_window:\n",
    "                for i in RESIDUES:\n",
    "                    self.df.loc[index ,i+'_lcscore'] = (len(low_complexity_list.loc[\n",
    "                        (low_complexity_list['bool'] == True) &\n",
    "                        (low_complexity_list['acid'] == i)])\n",
    "                    )\n",
    "                    self.df.loc[index ,i+'_lcfraction'] = (len(low_complexity_list.loc[\n",
    "                        (low_complexity_list['bool'] == True) & \n",
    "                        (low_complexity_list['acid'] == i)]) / len(lcs_acids)\n",
    "                    )\n",
    "        self.df['lcs_fractions'] = lcs_fractions\n",
    "        self.df['lcs_scores'] = lcs_scores\n",
    "        self.df['lcs_lowest_complexity'] = lcs_lowest_complexity\n",
    "        \n",
    "    def add_plaac(self):\n",
    "        plaac = CWD+'\\\\data\\\\plaac\\\\plaac_swissprot140219.tsv'\n",
    "        plaac = pd.read_csv(plaac, sep='\\t')\n",
    "        plaac[['database', 'accession', 'name']] = plaac['SEQid'].str.split('|',expand=True)\n",
    "        plaac = plaac.drop(['SEQid', 'database', 'name', 'PAPAaa', 'STARTaa', \n",
    "                            'ENDaa', 'COREaa', 'MW', 'MWstart', 'MWend', 'MWlen'], axis=1)\n",
    "        self.df = pd.merge(self.df, plaac, left_on='uniprot_id', right_on='accession')\n",
    "        self.df = self.df.drop('accession', axis=1)\n",
    "\n",
    "\n",
    "def convolve_signal(sig, window=25):\n",
    "    win = signal.hann(window)\n",
    "    sig = signal.convolve(sig, win, mode='same') / sum(win)\n",
    "    return sig\n",
    "\n",
    "\n",
    "def average(l):\n",
    "    return sum(l) / len(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def main(name, fasta_path, operating_system='Windows'):\n",
    "    \"\"\" Generates and saves a file which contains features of a protein sequence.\n",
    "    Parameters:\n",
    "        name: Name of the file.\n",
    "        fasta_path: Path of the fasta file which needs to be featured.\n",
    "        operating_system: String which indicates which operating system is used only 'Windows' available.\n",
    "    \"\"\"\n",
    "    data = MakeMatrix(fasta_path)   \n",
    "    now = datetime.datetime.now()\n",
    "    date = (str(now.day) + '-' + str(now.month)  + '-' +  str(now.year))\n",
    "    if operating_system == 'Windows':\n",
    "         data.df.to_pickle('data\\\\Dataframes\\\\'+ name + '_llps_f2f_' + date + '.pkl')\n",
    "    print('Generated file: ' + name + '_llps_f2f_' + date + '.pkl')\n",
    "    return data.df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Homo sapiens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "human_proteome_path = CWD+'\\\\data\\\\fasta\\\\human.fasta'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "500 fasta sequences per 1 minute on my: <i>AMD 965 AMD Phenom II X4 965 Black Edition</i>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14233\n",
      "0.35s self.fasta2df(dbfasta)\n",
      "46.58s self.amino_acid_analysis()\n",
      "794.81s self.idr_iupred()\n",
      "56.74s self.hydrophobic()\n",
      "72.65s self.add_iupred_features()\n",
      "111.64s self.add_hydrophobic_features()\n",
      "0.71s self.add_biochemical_combinations()\n",
      "95.51s self.add_lowcomplexity_features()\n",
      "Generated file: PDB_human_llps_f2f_20-12-2019.pkl\n"
     ]
    }
   ],
   "source": [
    "# Human\n",
    "df = main(name='PDB_human', fasta_path=human_proteome_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
